internal/indexmanagement/reconcile.go

```diff
@@ -64,80 +66,246 @@ type rolloverConditions struct {
+			ll.Error(err, "could not reconcile indexmanagement cronjob")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (imr *IndexManagementRequest) cullIndexManagement(mappings []apis.IndexManagementPolicyMappingSpec, policies apis.PolicyMap) {
+	if err := imr.removeCronJobsForMappings(mappings, policies); err != nil {
+		log.Error(err, "Unable to cull cronjobs")
+	}
+	mappingNames := sets.NewString()
 	for _, mapping := range mappings {
-		expected.Insert(fmt.Sprintf("%s-im-%s", cluster.Name, mapping.Name))
+		mappingNames.Insert(formatTemplateName(mapping.Name))
+	}
+
+	existing, err := imr.esClient.ListTemplates()
+	if err != nil {
+		log.Error(err, "Unable to list existing templates in order to reconcile stale ones")
+		return
+	}
+	difference := existing.Difference(mappingNames)
+
+	for _, template := range difference.List() {
+		if strings.HasPrefix(template, constants.OcpTemplatePrefix) {
+			if err := imr.esClient.DeleteIndexTemplate(template); err != nil {
+				log.Error(err, "Unable to delete stale template in order to reconcile", "template", template)
+			}
+		}
+	}
+}
+
+func (imr *IndexManagementRequest) initializeIndexIfNeeded(mapping apis.IndexManagementPolicyMappingSpec) error {
+	pattern := formatWriteAlias(mapping)
+	indices, err := imr.esClient.ListIndicesForAlias(pattern)
+	if err != nil {
+		return err
+	}
+	if len(indices) < 1 {
+		indexName := fmt.Sprintf("%s-000001", mapping.Name)
+		primaryShards := int32(elasticsearch.CalculatePrimaryCount(imr.cluster))
+		replicas := int32(elasticsearch.CalculateReplicaCount(imr.cluster))
+		index := esapi.NewIndex(indexName, primaryShards, replicas)
+		index.AddAlias(mapping.Name, false)
+		index.AddAlias(pattern, true)
+		for _, alias := range mapping.Aliases {
+			index.AddAlias(alias, false)
+		}
+		return imr.esClient.CreateIndex(indexName, index)
+	}
+	return nil
+}
+
+func formatTemplateName(name string) string {
+	return fmt.Sprintf("%s-%s", constants.OcpTemplatePrefix, name)
+}
+
+func formatWriteAlias(mapping apis.IndexManagementPolicyMappingSpec) string {
+	return fmt.Sprintf("%s-write", mapping.Name)
+}
+
+func (imr *IndexManagementRequest) createOrUpdateIndexTemplate(mapping apis.IndexManagementPolicyMappingSpec) error {
+	name := formatTemplateName(mapping.Name)
+	pattern := fmt.Sprintf("%s*", mapping.Name)
+	primaryShards := int32(elasticsearch.CalculatePrimaryCount(imr.cluster))
+	replicas := int32(elasticsearch.CalculateReplicaCount(imr.cluster))
+	aliases := append(mapping.Aliases, mapping.Name)
+	template := esapi.NewIndexTemplate(pattern, aliases, primaryShards, replicas)
+
+	// check to compare the current index templates vs what we just generated
+	templates, err := imr.esClient.GetIndexTemplates()
+	if err != nil {
+		return err
 	}
 
-	cronList := &batch.CronJobList{}
-	listOpts := []client.ListOption{
-		client.InNamespace(cluster.Namespace),
-		client.MatchingLabels(imLabels),
+	for templateName := range templates {
+		if templateName == name {
+			return nil
+		}
```

Comment:
what is this providing for us?

Reply:
TBH this was always there, the diff is a bit awkwardly rendered here. The move of this file info the `internal/indexmanagement` pkg yields into this diff.